include/seqan3/io/stream/parse_condition_detail.hpp

```diff
@@ -0,0 +1,490 @@
+
+/*!\brief A data structure that implements a subset of std::bitset as constexpr.
+ * \ingroup stream
+ * \tparam N                The number of bits.
+ */
+template <size_t N>
+class constexpr_pseudo_bitset : public std::array<bool, N>
+{
+private:
+    //!\brief The base type.
+    using base_t = std::array<bool, N>;
+public:
+    //!\brief Inherit constructors.
+    using base_t::base_t;
+
+    //!\brief Return a new bitset that is a logical disjunction of the two given ones.
+    constexpr constexpr_pseudo_bitset operator|(constexpr_pseudo_bitset rhs) const noexcept
+    {
+        for (size_t i = 0; i < N; ++i)
+            rhs[i] = rhs[i] || base_t::operator[](i);
+
+        return std::move(rhs);
+    }
+    //!\brief Return a new bitset with all bits flipped.
+    constexpr constexpr_pseudo_bitset operator~() const noexcept
+    {
+        constexpr_pseudo_bitset ret{};
+        for (size_t i = 0; i < N; ++i)
+            ret[i] = !base_t::operator[](i);
+
+        return ret;
+    }
+};
+
+// ----------------------------------------------------------------------------
+// condition_message_v
+// ----------------------------------------------------------------------------
+
+/*!\brief Defines a compound seqan3::constexpr_string consisting of all given conditions separated by the
+ *        operator-name `op`.
+ * \ingroup stream
+ * \tparam op               non-type template parameter specifying the separator character, e.g. '|'.
+ * \tparam condition_head_t The first condition type in the message. Ensures that there is at least one type.
+ * \tparam condition_ts     Remaining list of conditions separated by `op`.
+ * \relates seqan3::detail::parse_condition
+ */
+template <char op, typename condition_head_t, typename ...condition_ts>
+inline constexpr_string constexpr condition_message_v
+{
+    constexpr_string{"("} +
+    (condition_head_t::msg + ... +
+        (constexpr_string{" "} + constexpr_string{{op, op, '\0'}} + constexpr_string{" "} + condition_ts::msg)) +
+    constexpr_string{")"}
+};
+
+// ----------------------------------------------------------------------------
+// parse_condition_concept
+// ----------------------------------------------------------------------------
+
+//!\cond
+template <typename condition_t>
+class parse_condition_base;
+//!\endcond
+
+/*!\interface seqan3::detail::parse_condition_concept <>
+ * \brief An internal concept to check if an object fulfills the requirements of a seqan3::detail::parse_condition.
+ * \ingroup stream
+ *
+ * \details
+ *
+ * The must be invocable with an seqan3::char_adaptation_concept type and supply a static constexpr `msg` member of type
+ * seqan3::constexpr_string.
+ */
+//!\cond
+template <typename condition_t>
+concept parse_condition_concept = requires
+{
+    requires std::Predicate<std::remove_reference_t<condition_t>, char>;
+    requires std::is_base_of_v<parse_condition_base<remove_cvref_t<condition_t>>,
```

Comment:
so we have the concepts `is_type_specialization_of` and `is_base_of` isn't one superfluous?

Reply:
no, one checks type-specialisation and the other inheritance :wink: